### PR TITLE
[BugFix] Possible dead lock in tablet clone

### LINKED ISSUE #5646: possible dead lock in tablet clone
The `TabletSchemaMap::emplace` holds the lock of specific shard.
https://github.com/StarRocks/starrocks/blob/d50693ba41a6149067777de434e41e677ecf2224/be/src/storage/tablet_schema_map.cpp#L39-L42

The `TabletSchema` would be destructed when the smart pointer is reset and the origin object has no reference at all.
https://github.com/StarRocks/starrocks/blob/d50693ba41a6149067777de434e41e677ecf2224/be/src/storage/tablet_schema_map.cpp#L57

The desctructor of `TabletSchema` also needs the shard lock, which leads recursive locking.
https://github.com/StarRocks/starrocks/blob/d50693ba41a6149067777de434e41e677ecf2224/be/src/storage/tablet_schema_map.cpp#L67-L69

### Steps to reproduce the behavior (Required)
It's not easy to reproduce. The steps could refer #4514.

### Expected behavior (Required)
BE works well.
### Real behavior (Required)
BE gets stuck and dose not response to any request. 
image

### StarRocks version (Required)
main, branch-2.2, branch-2.1, branch-2.0

### PR DESCRIPTION
**Pull request by sevev**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #5646

## Problem Summary(Required) ：

We use a shared tablet schema to save memory usage, but the premise is that we need to ensure that no two different TabletSchemaPBs have the same `id`, so we will check the consistency of `unique_id` when two `TableSchemaPB` has the same `id`.

However, there may raise a deadlock if we do the check in `TabletSchemaMap::emplace` as the scenario below：
1. Thead1 holds a shared schema in which `id` is x.
2. Thead2  calls`TabletSchemaMap::emplace` with a `TabletSchemaPB` whose `id` is x, and it will hold the lock of a specific shard
3. The consistency check of `unique_id` failed, and we will create a new `schema`.
4. Thead1 releases the schema and the use_count of the old schema is 1, the old schema is held by Thread2.
5. Thead2 will deconstruct the old schema, and it needs to hold  the lock of a specific shard which will cause a deadlock

To avoid the deadlock, we will release the shard lock after finish the consistency check, and the old schema will be deconstruct at the end of this function if needed.

### CODE REVIEW

File: be/src/storage/tablet_meta.cpp [hunk 1]
```diff
@@ -472,6 +489,12 @@ void TabletMeta::init_from_pb(TabletMetaPB* ptablet_meta_pb) {
     if (tablet_meta_pb.schema().has_id() && tablet_meta_pb.schema().id() != TabletSchema::invalid_id()) {
         // Does not collect the memory usage of |_schema|.
         _schema = GlobalTabletSchemaMap::Instance()->emplace(tablet_meta_pb.schema()).first;
+        auto [_schema, inserted] = GlobalTabletSchemaMap::Instance()->emplace(tablet_meta_pb.schema());
```
Comment (rickif) [inline be/src/storage/tablet_meta.cpp:492]: I think it's better to add some explanation here.

[the remark this answers is not in the document]
Comment (sevev) [reply be/src/storage/tablet_meta.cpp]: > Why call this two times?

Line 491 should be deleted，changed in https://github.com/StarRocks/starrocks/pull/5661/commits/96693f9b1e8f32b3e1041c0ee3363c1007ea22f3

**Review by imay — DISMISSED** [review]

**Review by rickif — DISMISSED** [review]

**Comment by sevev** [comment]
@imay @sduzh @chaoyli update according comments.

**Comment by sevev** [comment]
update according to comments
Avoid the dead lock in `TabletSchemaMap::emplace` and execute the consistency check in this function to simplify the use of this function. PTAL
@imay @sduzh @chaoyli

**Review by sduzh — DISMISSED** [review]

File: be/src/storage/tablet_schema_map.cpp [hunk 2]
```diff
@@ -39,28 +39,43 @@ std::pair<TabletSchemaMap::TabletSchemaPtr, bool> TabletSchemaMap::emplace(const
...
-            it->second = std::weak_ptr<const TabletSchema>(ptr);
-            return std::make_pair(ptr, true);
+    // |result|: return value
+    // |ptr|: the shared_ptr whose id is equal to `id` in map
+    // |insert|: insert into map or not
+    // We use shared schema to save mem usage, but the premise is that we need to ensure that no two different
+    // TabletSchemaPBs have the same id. But we can't guarantee it after schema change so far, so we should
+    // check the consistent of tablet schema.
+    // If check failed, we will create a new tablet_schema as return value, but we must hold the original schema
+    // in map until the shard lock is release. If not, we may be deconstruct the original scheam which will cause
```
Comment (chaoyli) [inline be/src/storage/tablet_schema_map.cpp:49]: spell error.
`schema`

File: be/src/storage/tablet_schema_map.cpp [hunk 3]
```diff
@@ -39,28 +39,43 @@ std::pair<TabletSchemaMap::TabletSchemaPtr, bool> TabletSchemaMap::emplace(const
...
+            result = TabletSchema::create(_mem_tracker, schema_pb, this);
+            shard->map.emplace(id, result);
+            insert = true;
         } else {
-            if (UNLIKELY(!check_schema_unique_id(schema_pb, ptr))) {
-                ptr = TabletSchema::create(_mem_tracker, schema_pb, this);
-                it->second = std::weak_ptr<const TabletSchema>(ptr);
-                return std::make_pair(ptr, true);
+            ptr = it->second.lock();
+            if (UNLIKELY(!ptr)) {
```
Comment (chaoyli) [inline be/src/storage/tablet_schema_map.cpp:63]: `if (LIKELY(ptr))`

File: be/src/storage/tablet_schema_map.cpp [hunk 4]
```diff
@@ -39,28 +39,43 @@ std::pair<TabletSchemaMap::TabletSchemaPtr, bool> TabletSchemaMap::emplace(const
...
+                result = TabletSchema::create(_mem_tracker, schema_pb, this);
+                it->second = std::weak_ptr<const TabletSchema>(result);
+                insert = true;
+            } else {
+                if (UNLIKELY(!check_schema_unique_id(schema_pb, ptr))) {
```
Comment (chaoyli) [inline be/src/storage/tablet_schema_map.cpp:68]: `if (LIKELY(check_schema_unique_id(schema_pb, ptr)))`

File: be/src/storage/tablet_schema_map.cpp [hunk 5]
```diff
@@ -39,28 +39,43 @@ std::pair<TabletSchemaMap::TabletSchemaPtr, bool> TabletSchemaMap::emplace(const
...
+                } else {
+                    result = ptr;
+                }
+                insert = false;
             }
-            return std::make_pair(ptr, false);
         }
     }
+
+    return std::make_pair(result, insert);
```
Comment (chaoyli) [inline be/src/storage/tablet_schema_map.cpp:78]: If the insert if false, how to use the variable?

Comment (sevev) [reply be/src/storage/tablet_schema_map.cpp:78]: > If the insert if false, how to use the variable?

How to use this variable has nothing to do with the value of insert, the value of insert only represents whether the value in the map is upserted

File: be/src/storage/tablet_schema_map.cpp (hunk 2 above)
Comment (sevev) [reply be/src/storage/tablet_schema_map.cpp]: > spell error. `schema`

done，change in https://github.com/StarRocks/starrocks/pull/5661/commits/3732414a0bcc4ac2cfd0daa98ef613be98f384ca

**Review by sduzh — APPROVED** [review]

**Comment by sevev** [comment]
run starrocks_be_unittest

**Review by chaoyli — APPROVED** [review]

**Pull request merged**